lib/tools/api_python_module.py

```diff
@@ -0,0 +1,191 @@
+            'type': user_type,
+            'first_name': first_name,
+            'last_name': last_name
+            }
+        if section_name != None:
+            params['section_name'] = section_name
+        if grace_credits != None:
+            params['grace_credits'] = grace_credits
+        return self.submit_request(params, '/api/users', 'POST')
+
+    def get_assignments(self):
+        """ (ApiInterface) -> list of dict
+        Return a list of all assignments.
+        """
+        params = None
+        response = self.submit_request(params, '/api/assignments.json', 'GET')
+        return ApiInterface.decode_response(response)
+
+    def get_groups(self, assignment_id):
+        """ (ApiInterface, int) -> list of dict
+        Return a list of all groups associated with the given assignment.
+        """
+        params = None
+        path = self.get_path(assignment_id) + '.json'
+        response = self.submit_request(params, path, 'GET')
+        return ApiInterface.decode_response(response)
+
+    def get_groups_by_name(self, assignment_id):
+        """ (ApiInterface, int) -> dict of str:int
+        Return a dictionary mapping group names to group ids.
+        """
+        params = None
+        path = self.get_path(assignment_id) + '/group_ids_by_name.json'
+        response = self.submit_request(params, path, 'GET')
+        return ApiInterface.decode_response(response)
+
+    def upload_test_results(self, assignment_id, group_id, title, contents):
+        """ (ApiInterface, int, int, string, string) -> list of str
+        Upload test results to Markus.
+
+        Keyword arguments:
+        assignment_id -- the assignment's id
+        group_id      -- the id of the group to which we are uploading
+        title         -- the file name that will be displayed
+        contents      -- what will be in the file
+        """
+        params = {}
+        params['filename'] = title
+        params['file_content'] = contents
+        path = self.get_path(assignment_id, group_id) + 'test_results.xml'
+        return self.submit_request(params, path, 'POST')
+
+    def update_marks_single_group(self, criteria_mark_map,
+                                  assignment_id, group_id):
+        """ (ApiInterface, dict, int, int) -> list of str
+        Update the marks of a single group. 
+        Only the marks specified in criteria_mark_map will be changed.
+        To set a mark to unmarked, use 'nil' as it's value.
+        Otherwise, marks must have valid numeric types (floats or ints).
+        Criteria are specified by their title. Titles must be formatted
+        exactly as they appear in the MarkUs GUI, punctuation included.
+
+        Keyword arguments:
+        criteria_mark_map --  maps criteria to the desired grade.
+        assignment_id -- the assignment's id
+        group_id      -- the id of the group whose marks we are updating
+        """
+        params = criteria_mark_map
```

Comment:
I know this is redundant, but maintains the format used by each method for specifying the params first.